sound/soc/sof/ipc4-pcm.c

```diff
@@ -62,10 +62,37 @@ int sof_ipc4_set_pipeline_state(struct snd_sof_dev *sdev, u32 instance_id, u32 s
 }
 EXPORT_SYMBOL(sof_ipc4_set_pipeline_state);
 
+static void sof_ipc4_add_pipeline_by_priority(struct ipc4_pipeline_set_state_data *trigger_list,
+					      struct snd_sof_widget *pipe_widget,
+					      s8 *pipe_priority, bool ascend)
+{
+	struct sof_ipc4_pipeline *pipeline = pipe_widget->private;
+	int i, j;
+
+	for (i = 0; i < trigger_list->count; i++) {
+		/* add pipeline from low priority to high */
+		if (ascend && pipeline->priority < pipe_priority[i])
+			break;
+		/* add pipeline from high priority to low */
+		else if (!ascend && pipeline->priority > pipe_priority[i])
+			break;
+	}
+
+	for (j = trigger_list->count - 1; j >= i; j--) {
+		trigger_list->pipeline_instance_ids[j + 1] = trigger_list->pipeline_instance_ids[j];
+		pipe_priority[j + 1] = pipe_priority[j];
```

Comment:
`trigger_list->count - 1 + 1 == trigger_list->count`
`trigger_list->pipeline_instance_ids[trigger_list->count]` and `pipe_priority[trigger_list->count]` is out of bounds access?

Reply:
i had the same suspicion when I looked at the code. But the array size is pipeline_list->count  not trigger_list->count. So I think it is safe here

Reply:
At the end we are not going to put all pipelines in the trigger list? Right, when we add the last pipeline the `trigger_list->count` is `pipeline_list->count`, I guess it is OK, but super confusing.